src/net_processing.cpp

```diff
@@ -956,14 +1025,12 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const CValidationSta
     const uint256 hash(block.GetHash());
     std::map<uint256, std::pair<NodeId, bool>>::iterator it = mapBlockSource.find(hash);
 
-    int nDoS = 0;
-    if (state.IsInvalid(nDoS)) {
+    if (state.IsInvalid()) {
         // Don't send reject message with code 0 or an internal reject code.
         if (it != mapBlockSource.end() && State(it->second.first) && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) {
             CBlockReject reject = {(unsigned char)state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), hash};
             State(it->second.first)->rejects.push_back(reject);
-            if (nDoS > 0 && it->second.second)
-                Misbehaving(it->second.first, nDoS);
+            MaybePunishNode(it->second.first, state, !it->second.second);
```

Comment:
Happy to do it in a followup, this is already enough code and I don't want to risk a bug if it sits for another 6 months and requires rebase.

Reply:
Yes - can be done in a follow up, but in the meantime:

` MaybePunishNode(it->second.first, state, !it->second.second);`

is almost completely inscrutable!

Reply:
In fairness, it was this way when he found it. :P Def worthy of cleanup.

Reply:
FYI, a more self-documenting alternative looks something like:
```c++
if (state.IsInvalid()) {      
    // Don't send reject message with code 0 or an internal reject code.
    NodeId node_id;
    bool punish_if_block_invalid;
    std::tie(node_id, punish_if_block_invalid) = it->second;
    if (it != mapBlockSource.end() && State(node_id) && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) {
        CBlockReject reject = {(unsigned char)state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), hash};
        State(node_id)->rejects.push_back(reject);
        MaybePunishNode(node_id, state, !punish_if_block_invalid);
    }
}
```
Note you can early return for the `== end()` case here because it's present in every case.